sycl/source/handler.cpp

```diff
@@ -0,0 +1,245 @@
+    CommandGroup.reset(new detail::CGExecKernel(
+        std::move(MNDRDesc), std::move(MHostKernel), std::move(MSyclKernel),
+        std::move(MArgsStorage), std::move(MAccStorage),
+        std::move(MSharedPtrStorage), std::move(MRequirements),
+        std::move(MEvents), std::move(MArgs), std::move(MKernelName),
+        std::move(MOSModuleHandle), std::move(MStreamStorage), MCGType));
+    break;
+  }
+  case detail::CG::COPY_ACC_TO_PTR:
+  case detail::CG::COPY_PTR_TO_ACC:
+  case detail::CG::COPY_ACC_TO_ACC:
+    CommandGroup.reset(
+        new detail::CGCopy(MCGType, MSrcPtr, MDstPtr, std::move(MArgsStorage),
+                           std::move(MAccStorage), std::move(MSharedPtrStorage),
+                           std::move(MRequirements), std::move(MEvents)));
+    break;
+  case detail::CG::FILL:
+    CommandGroup.reset(new detail::CGFill(
+        std::move(MPattern), MDstPtr, std::move(MArgsStorage),
+        std::move(MAccStorage), std::move(MSharedPtrStorage),
+        std::move(MRequirements), std::move(MEvents)));
+    break;
+  case detail::CG::UPDATE_HOST:
+    CommandGroup.reset(new detail::CGUpdateHost(
+        MDstPtr, std::move(MArgsStorage), std::move(MAccStorage),
+        std::move(MSharedPtrStorage), std::move(MRequirements),
+        std::move(MEvents)));
+    break;
+  case detail::CG::COPY_USM:
+    CommandGroup.reset(new detail::CGCopyUSM(
+        MSrcPtr, MDstPtr, MLength, std::move(MArgsStorage),
+        std::move(MAccStorage), std::move(MSharedPtrStorage),
+        std::move(MRequirements), std::move(MEvents)));
+    break;
+  case detail::CG::FILL_USM:
+    CommandGroup.reset(new detail::CGFillUSM(
+        std::move(MPattern), MDstPtr, MLength, std::move(MArgsStorage),
+        std::move(MAccStorage), std::move(MSharedPtrStorage),
+        std::move(MRequirements), std::move(MEvents)));
+    break;
+  case detail::CG::PREFETCH_USM:
+    CommandGroup.reset(new detail::CGPrefetchUSM(
+        MDstPtr, MLength, std::move(MArgsStorage), std::move(MAccStorage),
+        std::move(MSharedPtrStorage), std::move(MRequirements),
+        std::move(MEvents)));
+    break;
+  case detail::CG::NONE:
+    throw runtime_error("Command group submitted without a kernel or a "
+                        "explicit memory operation.");
+  default:
+    throw runtime_error("Unhandled type of command group");
+  }
+
+  detail::EventImplPtr Event = detail::Scheduler::getInstance().addCG(
+      std::move(CommandGroup), std::move(MQueue));
+
+  EventRet = detail::createSyclObjFromImpl<event>(Event);
+  return EventRet;
+}
+
+void handler::processArg(void *Ptr, const detail::kernel_param_kind_t &Kind,
+                         const int Size, const size_t Index, size_t &IndexShift,
+                         bool IsKernelCreatedFromSource) {
+  const auto kind_std_layout = detail::kernel_param_kind_t::kind_std_layout;
+  const auto kind_accessor = detail::kernel_param_kind_t::kind_accessor;
```

Comment:
Where is `kind_accessor`/`kind_pointer` used?

Also I suggest using names not matching enum values to simplify reading. It looks like you mixed-up constants and enum values.

Reply:
@bader those are used at lines 98 and 177. I have added postfix `_v` to avoid confusion.